hoot-josm/src/main/cpp/hoot/core/josm/ops/JosmMapCleaner.cpp

```diff
@@ -0,0 +1,286 @@
+
+    LOG_DEBUG("Reading cleaned map from " << tempOutputPath << "...");
+    OsmMapPtr cleanedMap(new OsmMap());
+    OsmXmlReader reader;
+    reader.setUseDataSourceIds(true);
+    reader.setUseFileStatus(true);
+    reader.open(tempOutputPath);
+    reader.read(tempOutputPath, cleanedMap);
+    return cleanedMap;
+  }
+  else
+  {
+    // pass map as string and get it back as a string
+    return
+      OsmXmlReader::fromXml(
+        _clean(_josmValidators, OsmXmlWriter::toString(inputMap, false), _addDetailTags).trimmed(),
+        true, true, false, true);
+  }
+}
+
+QString JosmMapCleaner::_clean(
+  const QStringList& validators, const QString& map, const bool addDetailTags)
+{
+  // JNI sig format: (input params...)return type
+  jstring cleanedMapResultStr =
+    (jstring)_javaEnv->CallObjectMethod(
+      _josmInterface,
+      // Java sig:
+      //  String validate(
+      //    List<String> validators, String elementsXml, boolean cleanValidated, boolean addTags)
+      _javaEnv->GetMethodID(
+        _josmInterfaceClass, "validate",
+        "(Ljava/util/List;Ljava/lang/String;ZZ)Ljava/lang/String;"),
+      JniConversion::toJavaStringList(_javaEnv, validators),
+      JniConversion::toJavaString(_javaEnv, map),
+      true,
+      addDetailTags);
+  JniUtils::checkForErrors(_javaEnv, "cleanFromMapString");
+  return JniConversion::fromJavaString(_javaEnv, cleanedMapResultStr);
+}
+
+void JosmMapCleaner::_clean(
+  const QStringList& validators, const QString& inputMapPath, const QString& outputMapPath,
+  const bool addDetailTags)
+{
+  // JNI sig format: (input params...)return type
+  _javaEnv->CallVoidMethod(
+    _josmInterface,
+    // Java sig:
+    //  void validate(
+    //    List<String> validators, String elementsFileInputPath, String elementsFileOutputPath,
+    //    boolean cleanValidated, boolean addTags)
+    _javaEnv->GetMethodID(
+      _josmInterfaceClass, "validate", "(Ljava/util/List;Ljava/lang/String;Ljava/lang/String;ZZ)V"),
+    JniConversion::toJavaStringList(_javaEnv, validators),
+    JniConversion::toJavaString(_javaEnv, inputMapPath),
+    JniConversion::toJavaString(_javaEnv, outputMapPath),
+    true,
+    addDetailTags);
+  JniUtils::checkForErrors(_javaEnv, "cleanFromMapFile");
+}
+
+void JosmMapCleaner::_getStats()
+{
+  LOG_DEBUG("Retrieving stats...");
+
+  // call back into the hoot-josm validator to get the stats after validation
+
+  JosmMapValidatorAbstract::_getStats();
+  _numElementsCleaned = _getNumElementsCleaned();
+  _deletedElementIds = _getDeletedElementIds();
+  _numFailingValidators = _getNumFailingValidators();
+  _numFailedCleaningOperations = _getNumFailedCleaningOperations();
+
+  // create a readable error summary
+
+  _errorSummary =
+    "Total JOSM validation errors: " + StringUtils::formatLargeNumber(_numValidationErrors) +
+    " found in " + StringUtils::formatLargeNumber(_numAffected) + " total features.\n";
```

Comment:
Would it make more sense to add each of these to `SingleStat` objects instead of the `errorSummary` string?

Reply:
good idea. i'll take a look at it.

Reply:
Tried it but started to get nasty due to the way I'm building some of those error count strings. Put in a todo to clean it up some more for later.